### PR TITLE
Add grunt-esdoc to tool set

### PR DESCRIPTION
**Pull request by zimmi88**
As part of 0.3.0, we'd like to find a solution for generating documentation from our code that developers can reference when working with the core code. Unfortunately, JSDoc doesn't work with ES6 modules yet. ESDoc, on the other hand, recognizes ES6 code and can integrate nicely with Grunt. As a plus, it also generates coverage reports. =)

This PR creates a new grunt task `grunt esdoc` that generates documentation at `tmp/esdoc`. Developers can read this documentation locally and we can copy this code over into gh-pages. Later on, we can explore rebuilding and checking in this documentation as part of the release tasks.

Caveat: ESDoc can't parse non-ES6 code. Though, given we intend to convert hopscotch.js into an ES6 module, this shouldn't be a problem.

We're not at 100% documentation coverage with this PR, but I propose we pull in these commits sooner rather than later to avoid diverging too far away from the gh72 branch.

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -312,6 +312,14 @@ module.exports = function (grunt) {
         configFile: 'eslint.json'
       },
       target: ['<%=paths.source%>/es/**/*.js', '<%=paths.source%>/es/**/*.js']
+    },
+    esdoc: {
+      dev: {
+        options: {
+          "source": "./src/es",
+          "destination": "./tmp/esdoc"
```
Comment (kate2753) [inline Gruntfile.js]: Could we use variables for folders here, like this:
```
options:  {
  "source" : "<%=paths.source%>/es",
  "destination" : "<%=paths.build%>/esdoc"
}
```

File: src/es/hopscotch.js [hunk 2]
```diff
@@ -42,13 +42,19 @@ import * as Utils from './modules/utils.js';
...
   let calloutMan;
+  let tmplClosureOut = {};
   
   // Template includes, placed inside a closure to ensure we don't
   // end up declaring our shim globally.
   (function () {
     // @@include('../../src/tl/_template_headers.js') //
     // @@include('../../tmp/js/hopscotch_templates.js') //
-  }.call(TemplateManager));
+  }.call(tmplClosureOut));
```
Comment (kate2753) [inline src/es/hopscotch.js]: Clever :)

**Comment by kate2753** [comment]
This looks great! Thanks for taking time to update the docs!

File: Gruntfile.js (hunk 1 above)
Comment (zimmi88) [inline Gruntfile.js]: Fixed.

**Comment by h13i32maru** [comment]
Hi! I am developing ESDoc. Thanks for using it!

[The ESDoc Hosting Service](https://doc.esdoc.org) can host your documentation. It only support raw ``esdoc.config`` (not grunt task, gulp task and etc). Please consider it :smile: 

Feel free this comment.
Thanks.

**Pull request merged**